File: cpp/189_旋转数组_medium.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int N = nums.size();
        k %= N;
        reverse(nums, 0, N-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, N-1);
    }
    void reverse(vector<int> &nums, int start, int end){
        while(start<end){
            swap(nums[start], nums[end]);
            start += 1;
            end -= 1;
        }
    }
};

File: python/189_旋转数组_medium.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def reverse(start, end):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
        N = len(nums)
        k %= N
        reverse(0, N-1)
        reverse(0, k-1)
        reverse(k, N-1)